src/Lumper.UI/Services/BspService.cs

```diff
@@ -0,0 +1,327 @@
+            {
+                var outName = outFile is not null ? Path.GetFileName(outFile.Path.AbsolutePath) : FileName;
+                progressWindow = new IoProgressWindow {
+                    Title = $"Saving {outName} ({compressString})",
+                    Handler = handler
+                };
+                _ = progressWindow.ShowDialog(Program.Desktop.MainWindow);
+            }
+
+            await Observable.Start(
+                () => BspFile.SaveToFile(
+                    handler,
+                    outFile?.Path.LocalPath,
+                    compress: compress,
+                    makeBackup: MakeBackup),
+                RxApp.TaskpoolScheduler);
+
+            progressWindow?.Close();
+
+            if (handler.Cancelled)
+                return false;
+        }
+        catch (FileLoadException)
+        {
+            Logger.Warn("Failed to load new file, doing a full UI file load");
+            await Observable.Start(() => Load(outFile?.Path.LocalPath ?? FilePath!),
+                RxApp.TaskpoolScheduler);
+        }
+        catch (Exception ex)
+        {
+            Logger.Error(ex, "Failed to save file!");
+            IsLoading = false;
+            return false;
+        }
+
+        UpdateBspProperties();
+        IsLoading = false;
+        IsModified = false;
+        return true;
+    }
+
+    /// <summary>
+    /// Close the currently loaded BSP file
+    /// </summary>
+    public void Close()
+    {
+        BspFile = null; // This calls UpdateBspProperties
+        ResetLumpViewModels();
+
+        IsModified = false;
+        IsLoading = false;
+    }
+
+    /// <summary>
+    /// Writes a JSON summary of the file out to {filename}.json
+    /// </summary>
+    public void JsonDump()
+    {
+        var cts = new CancellationTokenSource();
+        var handler = new IoHandler(cts);
+        Observable.Start(
+            () => BspFile?.JsonDump(handler, sortLumps: false, sortProperties: false, ignoreOffset: false),
+            RxApp.TaskpoolScheduler);
+    }
+
+    /// <summary>
+    /// Mark the currently loaded BSP file as modified
+    /// </summary>
+    public void MarkAsModified()
+    {
+        if (!HasLoadedBsp || IsModified)
+            return;
+
+        IsModified = true;
+    }
+
+    public void ResetLumpViewModels()
+    {
+        ResetLumpViewModel(nameof(EntityLumpViewModel));
+        ResetLumpViewModel(nameof(PakfileLumpViewModel));
+    }
+
+    /// <summary>
+    /// Sets lump models to null, which will raise change notifs for any page that's using them.
+    /// If the new value is really null, there's no BSP loaded (see LazyLoadLump).
+    /// Even if there is one loaded, the getter will make a new instance of the viewmodel in question.
+    /// So whilst we're setting to null here, to consumers it appears that the models just got set
+    /// to a real, non-null value.
+    /// </summary>
+    public void ResetLumpViewModel(string type)
```

Comment:
`Type type` instead of `string type` and pass `MyType.GetType()`

Reply:
Can't use a switch in that case but okay